src/openms/include/OpenMS/ANALYSIS/ID/FragmentIndex.h

```diff
@@ -0,0 +1,146 @@
+// ETH Zurich, and Freie Universitaet Berlin 2002-2022.
+//
+// This software is released under a three-clause BSD license:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of any author or any participating institution
+//    may be used to endorse or promote products derived from this software
+//    without specific prior written permission.
+// For a full list of authors, refer to the file AUTHORS.
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Chris Bielow $
+// $Authors: Max Alcer, Heike Einsfeld $
+// --------------------------------------------------------------------------
+
+
+#pragma once
+
+#include <OpenMS/CHEMISTRY/AASequence.h>
+#include <OpenMS/DATASTRUCTURES/DefaultParamHandler.h>
+#include <OpenMS/FORMAT/FASTAFile.h>
+#include <OpenMS/KERNEL/MSExperiment.h>
+#include <OpenMS/KERNEL/Peak1D.h>
+
+#include <vector>
+
+namespace OpenMS
+{
+  //class MSExperiment;
+
+  /** @brief Creating a two-level tree from a given FASTAFile by generating the theoretical peptides and their b-y-ions and sorting
+    * them in the outer tree by fragment-m/z and the inner tree by precursor-m/z. Search function to find candidates to experimental MS2 sepctra
+    * (implementation inspired by https://lazear.github.io/sage/)
+    * @ingroup ID
+   */
+class OPENMS_DLLAPI FragmentIndex : public DefaultParamHandler
+{
+  public:
+
+  /** @brief Storing found candiates from search 
+   */
+  struct Candidate
+  {
+    size_t peptide_start; ///< Start position of peptide in protein
+    size_t peptide_end; ///< End position of peptide in protein
+    size_t protein_index; ///< Index to Entry of std::vector<FASTAFile::FASTAEntry>
```

Comment:
do you also need to store a size_t modification_index ? 
e.g.: let's say you have a peptide TMMMTR
and modification variants
TM(Oxidation)MMTR
TMM(Oxidation)MTR
TMMM(Oxidation)TR
you might need to be able to reconstruct that the candidate refers to e.g. the second modification (isoform) of the peptide

Reply:
yes we do. we are just not yet shure how. An Index is not enough. The type must be stored too

Reply:
In principle, only one index is needed as modification isoforms are generated in the same order.
e.g. with multiple different modifications, you can still unambiguously use an index to indicate which isoform you mean
TM(Oxidation)STR
TMS(Phospho)TR
TM(Oxidation)S(Phospho)TR